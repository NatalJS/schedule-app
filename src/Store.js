import React from 'react';
import get from 'lodash/get';
import defaultTo from 'lodash/defaultTo';
import every from 'lodash/every';
import mapValues from 'lodash/mapValues';

class Store extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      ...this.reduceCalendarData(props.data),
      searchFilter: '',
      isShowingAdvancedFilters: false
    }

    this.actions = {
      onCategoryFilterChange: this.onFilterChange.bind(this, 'categoryFilter'),
      onTypeFilterChange: this.onFilterChange.bind(this, 'typeFilter'),
      filterEvents: this.filterEvents.bind(this),
      toggleAdvancedFilters: this.toggleAdvancedFilters.bind(this),
      onSearchFilterChange: this.onSearchFilterChange.bind(this),
      checkSearchMatch: this.checkSearchMatch.bind(this),
    }
  }

  reduceCalendarData(data) {
    const days = {'22': []};
    const eventTypes = ['Eventos Fixos'];
    const talksCategories = [];

    data.items.forEach(event => {
      const startDateTime = get(event, 'start.dateTime');
      if (!startDateTime) {
        return;
      }
      const dayOfEvent = new Date(startDateTime).getDate();
      if ([22].includes(dayOfEvent)) {
        return;
      }
      if (!days[dayOfEvent]) days[dayOfEvent] = [];

      const pybrEvent = {
        id: event.id,
        date: new Date(startDateTime),
        summary: event.summary,
        details: {
          eventType: 'Eventos Fixos'
        }
      }

      if (event.description) {
        const [ name, title, eventType, category, duration ] = event.description.split('|').map(i => i.trim());
        pybrEvent.details = {
          name,
          title,
          eventType,
          category,
          duration
        };
        if (!eventTypes.includes(eventType)) eventTypes.push(eventType);
        if (category && !talksCategories.includes(category)) talksCategories.push(category);
      }
      const eventsOnSameTime = days[dayOfEvent].find(h => h.date.getTime() == pybrEvent.date.getTime());
      if (!eventsOnSameTime) {
        days[dayOfEvent].push({
          date: pybrEvent.date,
          events: [pybrEvent]
        })
      } else {
        eventsOnSameTime.events.push(pybrEvent);
      }
    });
    for (const day in days) {
      days[day].sort(this.sortByDate);
    }
    return {
      days,
      eventTypes,
      talksCategories,
      categoryFilter: [ ...talksCategories ],
      typeFilter: [ ...eventTypes ]
    }
  }

  sortByDate(eventA, eventB) {
    if (eventA.date == eventB.date) {
      return 0;
    }

    return eventA.date > eventB.date ? 1 : -1;
  }

  onFilterChange(filterType, filter) {
    const state = this.state;
    if (state[filterType].includes(filter)) {
      this.setState((state) => ({
        ...state,
        [filterType]: state[filterType].filter(f => f !== filter)
      }))
    } else {
      this.setState((state) => ({
        ...state,
        [filterType]: [ ...state[filterType], filter ]
      }))
    }
  }

  onSearchFilterChange(e) {
    this.setState({
      searchFilter: e.target.value
    })
  }

  toggleAdvancedFilters() {
    this.setState(state => ({
      ...state,
      isShowingAdvancedFilters: !state.isShowingAdvancedFilters
    }))
  }

  checkSearchMatch(event) {
    const searchRegex = new RegExp(this.state.searchFilter.toLowerCase(), 'i');
    return defaultTo(get(event, 'details.title'), '').match(searchRegex)
      || defaultTo(get(event, 'details.name'), '').match(searchRegex)
      || defaultTo(get(event, 'summary'), '').match(searchRegex);
  }

  filterEvents(acc, { date, events }) {
    const filteredEvents = events.filter(event => (
      this.state.typeFilter.includes(event.details.eventType)
        && (!event.details.category || this.state.categoryFilter.includes(event.details.category))
        && (!this.state.searchFilter || this.checkSearchMatch(event))
    ));
    if (filteredEvents.length)
      return [ ...acc, { date, events: filteredEvents } ];
    return acc;
  }

  render() {
    const { days} = this.state;
    const filteredDays = mapValues(days, day => day.reduce(this.actions.filterEvents, []));
    const isListEmpty = every(filteredDays, day => !day.length);
    return this.props.children({
      ...this.state,
      isListEmpty,
      days: filteredDays,
      actions: this.actions
    });
  }
}

export default Store;
