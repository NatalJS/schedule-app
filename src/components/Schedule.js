import React from 'react';
import map from 'lodash/map';
import reduce from 'lodash/reduce';
import classNames from 'classnames';
import { Sticky, StickyContainer } from 'react-sticky';
import DayMenu from './DayMenu';
import DaySeparator from './DaySeparator';
import ScrollNavigation from 'scroll-navigation-menu';
import Events from './Events';
import { getFormattedTime } from 'app/schedule-app/utils';
import { FilterBox, CategoryFilter, EventTypeFilter } from './filters';

class Schedule extends React.Component {
  renderDay(day, label) {
    if (day.length) {
      return (
        <div id={`day${label}`} >
          <DaySeparator day={label}/>
          {day.map(events => (
            <Events scheduleInDate={events} key={getFormattedTime(events.date)} />
          ))}
        </div>
      )
    }
  }

  componentDidMount() {
    // FIXME(ja0n): is this needed?
    const anchorsOffset = 0;

    this.anchors = new ScrollNavigation({
      offset: -anchorsOffset - 120
    });
    this.anchors.start();
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.store.days) {
      let items = 0;
      for (let key in this.props.store.days) {
        items += this.props.store.days[key].length;
      }

      if (items !== this.amountOfDays) {
        this.amountOfDays = items;
        this.anchors.scrollTo('.schedule-page', { minDuration: 0, maxDuration: 0 });
      }
    }
  }

  render() {
    const { store } = this.props;
    // FIXME(ja0n): is this needed?
    const stickyOffset = 0;

    return (
      <StickyContainer>
        <Sticky topOffset={-stickyOffset}>
          {({ style }) => (
            <div
              className="filters"
              style={{ ...style, top: stickyOffset  }}
            >
              <DayMenu days={store.days} />
              <FilterBox
                value={store.searchFilter}
                onClick={store.actions.toggleAdvancedFilters}
                onChange={store.actions.onSearchFilterChange}
                isPopoverOpened={store.isShowingAdvancedFilters}
                advancedFilters={
                  <div className="advanced-filters">
                    <h2>Filtrar por</h2>
                    <h3>Categoria</h3>
                    <CategoryFilter
                      categories={store.talksCategories}
                      filter={store.categoryFilter}
                      onChange={store.actions.onCategoryFilterChange}
                    />
                    <h3>Tipo</h3>
                    <EventTypeFilter
                      types={store.eventTypes}
                      filter={store.typeFilter}
                      onChange={store.actions.onTypeFilterChange}
                    />
                  </div>
                }
              />
            </div>
          )}
        </Sticky>
       {store.isListEmpty && <p className="empty-message">Nenhum evento encontrado.</p>}
       {map(store.days, (day, label) => (
          <React.Fragment key={label}>
            {this.renderDay(day, label)}
          </React.Fragment>
        ))}
      </StickyContainer>
    )
  }
}

export default Schedule;
