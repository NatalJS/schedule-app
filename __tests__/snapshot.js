import React from 'react';
import renderer from 'react-test-renderer';

import Store from '../src/Store';
import Schedule from '../src/components/Schedule';

const mockData = require('./mock.json');

test('Snapshot testing', () => {
  const data = mockData;
  const component = renderer.create(
    <Store data={data}>
      {store => <Schedule store={store} />}
    </Store>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
