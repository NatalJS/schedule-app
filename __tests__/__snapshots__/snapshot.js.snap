// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing 1`] = `
<div
  onScroll={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
>
  <div>
    <div />
    <div
      className="filters"
      style={
        Object {
          "top": 0,
        }
      }
    >
      <ul
        className="accordion-tabs schedule_category_days"
      >
        <li
          className="tab-header-and-content"
        >
          <a
            className="tab-link scroll"
            href="#day17"
          >
            17
          </a>
        </li>
        <li
          className="tab-header-and-content"
        >
          <a
            className="tab-link scroll"
            href="#day18"
          >
            18
          </a>
        </li>
        <li
          className="tab-header-and-content"
        >
          <a
            className="tab-link scroll"
            href="#day19"
          >
            19
          </a>
        </li>
        <li
          className="tab-header-and-content"
        >
          <a
            className="tab-link scroll"
            href="#day20"
          >
            20
          </a>
        </li>
        <li
          className="tab-header-and-content"
        >
          <a
            className="tab-link scroll"
            href="#day21"
          >
            21
          </a>
        </li>
      </ul>
      <div
        className="filter-box"
      >
        <form
          className="filters-search"
          onSubmit={[Function]}
        >
          <i
            className="material-icons"
          >
            search
          </i>
          <input
            onChange={[Function]}
            placeholder="Pesquisar palestra, autor..."
            value=""
          />
        </form>
        <button
          className="filters-button"
          onClick={[Function]}
        >
          <i
            className="material-icons"
          >
            filter_list
          </i>
        </button>
      </div>
    </div>
  </div>
  <div
    id="day17"
  >
    <div
      className="day-separator tab-link"
    >
      Dia 
      17
       – 
      <span>
        Tutoriais
      </span>
    </div>
    <article
      className="schedule_article"
    >
      <h5
        className="schedule_time"
      >
        12h00
      </h5>
      <div
        className="picture-container"
      >
        <div
          className="schedule_picture"
        />
      </div>
      <div
        className="row w-100"
      >
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12"
        >
          <h2
            className="schedule_name w-100"
          >
            Bar
          </h2>
        </div>
      </div>
    </article>
  </div>
  <div
    id="day18"
  >
    <div
      className="day-separator tab-link"
    >
      Dia 
      18
       – 
      <span>
        Tutoriais
      </span>
    </div>
    <article
      className="schedule_article"
    >
      <h5
        className="schedule_time"
      >
        12h00
      </h5>
      <div
        className="picture-container"
      >
        <div
          className="schedule_picture"
        />
      </div>
      <div
        className="row w-100"
      >
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12"
        >
          <h2
            className="schedule_name w-100"
          >
            Bar
          </h2>
        </div>
      </div>
    </article>
  </div>
  <div
    id="day19"
  >
    <div
      className="day-separator tab-link"
    >
      Dia 
      19
       – 
      <span>
        Palestras
      </span>
    </div>
    <article
      className="schedule_article"
    >
      <h5
        className="schedule_time"
      >
        10h05
      </h5>
      <div
        className="picture-container"
      >
        <div
          className="schedule_picture"
        />
      </div>
      <div
        className="row w-100"
      >
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12 schedule-highlight"
        >
          <h2
            className="schedule_name"
          >
            2 anos de trabalho remoto - o lado bom e os problemas
          </h2>
          <h3
            className="schedule_speaker"
          >
            John Doe
          </h3>
          <h4
            className="schedule_location"
          >
            <i
              className="material-icons"
            >
              location_on
            </i>
            Sala Jerimum
          </h4>
        </div>
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12 schedule-highlight"
        >
          <h2
            className="schedule_name"
          >
            O casamento sagrado entre o Pipfile, pip e Virtualenv
          </h2>
          <h3
            className="schedule_speaker"
          >
            John Doe
          </h3>
          <h4
            className="schedule_location"
          >
            <i
              className="material-icons"
            >
              location_on
            </i>
            Sala Macaxeira
          </h4>
        </div>
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12 schedule-highlight"
        >
          <h2
            className="schedule_name"
          >
            Perdendo o medo da Web com Django
          </h2>
          <h3
            className="schedule_speaker"
          >
            John Doe
          </h3>
          <h4
            className="schedule_location"
          >
            <i
              className="material-icons"
            >
              location_on
            </i>
            Sala Rapadura
          </h4>
        </div>
      </div>
    </article>
    <article
      className="schedule_article"
    >
      <h5
        className="schedule_time"
      >
        10h35
      </h5>
      <div
        className="picture-container"
      >
        <div
          className="schedule_picture"
        />
      </div>
      <div
        className="row w-100"
      >
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12 schedule-highlight"
        >
          <h2
            className="schedule_name"
          >
            Ansible 101
          </h2>
          <h3
            className="schedule_speaker"
          >
            John Doe
          </h3>
          <h4
            className="schedule_location"
          >
            <i
              className="material-icons"
            >
              location_on
            </i>
            Sala Macaxeira
          </h4>
        </div>
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12 schedule-highlight"
        >
          <h2
            className="schedule_name"
          >
            Criando clients inteligentes com o simple-rest-client
          </h2>
          <h3
            className="schedule_speaker"
          >
            John Doe
          </h3>
          <h4
            className="schedule_location"
          >
            <i
              className="material-icons"
            >
              location_on
            </i>
            Sala Rapadura
          </h4>
        </div>
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12 schedule-highlight"
        >
          <h2
            className="schedule_name"
          >
            Web crawling e scraping com Scrapy e Scrapy Cloud
          </h2>
          <h3
            className="schedule_speaker"
          >
            John Doe
          </h3>
          <h4
            className="schedule_location"
          >
            <i
              className="material-icons"
            >
              location_on
            </i>
            Sala Tapioca
          </h4>
        </div>
      </div>
    </article>
    <article
      className="schedule_article"
    >
      <h5
        className="schedule_time"
      >
        12h05
      </h5>
      <div
        className="picture-container"
      >
        <div
          className="schedule_picture"
        />
      </div>
      <div
        className="row w-100"
      >
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12"
        >
          <h2
            className="schedule_name w-100"
          >
            Bar
          </h2>
        </div>
      </div>
    </article>
  </div>
  <div
    id="day20"
  >
    <div
      className="day-separator tab-link"
    >
      Dia 
      20
       – 
      <span>
        Palestras
      </span>
    </div>
    <article
      className="schedule_article"
    >
      <h5
        className="schedule_time"
      >
        12h05
      </h5>
      <div
        className="picture-container"
      >
        <div
          className="schedule_picture"
        />
      </div>
      <div
        className="row w-100"
      >
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12"
        >
          <h2
            className="schedule_name w-100"
          >
            Bar
          </h2>
        </div>
      </div>
    </article>
  </div>
  <div
    id="day21"
  >
    <div
      className="day-separator tab-link"
    >
      Dia 
      21
       – 
      <span>
        Palestras
      </span>
    </div>
    <article
      className="schedule_article"
    >
      <h5
        className="schedule_time"
      >
        10h05
      </h5>
      <div
        className="picture-container"
      >
        <div
          className="schedule_picture"
        />
      </div>
      <div
        className="row w-100"
      >
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12 schedule-highlight"
        >
          <h2
            className="schedule_name"
          >
            Que belo dia para a ciência!
          </h2>
          <h3
            className="schedule_speaker"
          >
            John Doe
          </h3>
          <h4
            className="schedule_location"
          >
            <i
              className="material-icons"
            >
              location_on
            </i>
            Sala Tapioca
          </h4>
        </div>
      </div>
    </article>
    <article
      className="schedule_article"
    >
      <h5
        className="schedule_time"
      >
        12h05
      </h5>
      <div
        className="picture-container"
      >
        <div
          className="schedule_picture"
        />
      </div>
      <div
        className="row w-100"
      >
        <div
          className="schedule_info col-xl-3 col-lg-6 col-sm-12"
        >
          <h2
            className="schedule_name w-100"
          >
            Bar
          </h2>
        </div>
      </div>
    </article>
  </div>
</div>
`;
